### PR TITLE
Adds messaging sampler

### PR DESCRIPTION
**Pull request by adriancole**
this is pulling out #914 like we did for RPC

- [x] kafka-clients
- [x] jms
- [x] spring-rabbit

### CODE REVIEW

File: instrumentation/messaging/src/main/java/brave/messaging/MessagingRequest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+   */
+  public abstract Object unwrap();
+
+  /**
+   * The unqualified, case-sensitive semantic message operation name. The currently defined names
+   * are "send", "receive", "bulk-send" and "bulk-receive".
+   *
+   * <p>Examples:
+   * <pre><ul>
+   *   <li>Amazon SQS - {@code AmazonSQS.sendMessageBatch()} returns "bulk-send" as it can send many messages</li>
```
Comment (adriancole) [inline instrumentation/messaging/src/main/java/brave/messaging/MessagingRequest.java]: ps I'm still on the fence about bulk-X vs X-bulk :P

Comment (devinsba) [reply instrumentation/messaging/src/main/java/brave/messaging/MessagingRequest.java]: I'm thinking `send-bulk`

**Comment by jorgheymans** [comment]
stepping back a bit, why conveying bulk/batch notion at all in the
operation name? It's at most a hint that a batch send or retrieve could
have happened. Trying to see if i would care enough to see this level of
detail in a trace, and if it could not lead to false assumptions or
reasoning ie 'receive-batch' in the trace but the kafka consumer is
actually configured with max.poll.records=1 ...

**Comment by adriancole** [comment]
> stepping back a bit, why conveying bulk/batch notion at all in the
operation name? It's at most a hint that a batch send or retrieve could
have happened. Trying to see if i would care enough to see this level of
detail in a trace, and if it could not lead to false assumptions or
reasoning ie 'receive-batch' in the trace but the kafka consumer is
actually configured with max.poll.records=1 ...

For now, I agree. I'll move your comments into RATIONALE.md and we can change the impl easier to include batch later vs remove it. If we did that, we'd update RATIONALE with why

**Comment by adriancole** [comment]
updated with rationale

**Comment by adriancole** [comment]
@jeqo I added a rationale we kindof always forget about clearing headers. Notably, I've something in here about how to handle JMS which I think is more elegant than before.

**Comment by adriancole** [comment]
added kafka sampling impl finally. took a while to figure out what the best way to do this is..

**Comment by adriancole** [comment]
OK I integrated the header work from prior pull requests into kafka. With a look, I'll proceed to JMS

File: instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java [hunk 2]
```diff
@@ -110,8 +109,9 @@ public KafkaTracing build() {
     this.remoteServiceName = builder.remoteServiceName;
   }
 
-  public Tracing tracing() {
-    return tracing;
+  /** @since 5.9 exposed for Kafka Streams tracing. */
```
Comment (adriancole) [inline instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java]: @simondean fyi

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 3]
```diff
@@ -44,25 +49,27 @@
...
-    return new KafkaStreamsTracing.Builder(KafkaTracing.create(tracing)).build();
+    return create(KafkaTracing.create(tracing));
   }
 
+  /** @since 5.9 */
   public static KafkaStreamsTracing create(KafkaTracing kafkaTracing) {
-    return new KafkaStreamsTracing.Builder(kafkaTracing).build();
-  }
-
-  public KafkaTracing kafkaTracing() {
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]: no use of this so I removed this. In general, we expose things public upon use case, not before.

File: instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java [hunk 4]
```diff
@@ -134,27 +140,44 @@ public Tracing tracing() {
...
+  /** Creates a potentially noop remote span representing this request */
+  Span nextMessagingSpan(
+    SamplerFunction<MessagingRequest> sampler,
+    MessagingRequest request,
+    TraceContextOrSamplingFlags extracted
+  ) {
+    Boolean sampled = extracted.sampled();
+    // only recreate the context if the messaging sampler made a decision
+    if (sampled == null && (sampled = sampler.trySample(request)) != null) {
+      extracted = extracted.sampled(sampled.booleanValue());
```
Comment (jeqo) [inline instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java]: sweet! so this is the place where sampling will be overwritten if `samplerFunction` is provided, effectively solving the issue I post a few days ago in gitter :)

File: instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java [hunk 5]
```diff
@@ -81,33 +85,35 @@ public KafkaTracing build() {
...
+  final SamplerFunction<MessagingRequest> producerSampler, consumerSampler;
   final Set<String> propagationKeys;
   final String remoteServiceName;
 
   KafkaTracing(Builder builder) { // intentionally hidden constructor
-    this.tracing = builder.tracing;
-    this.extractor = tracing.propagation().extractor(KafkaPropagation.GETTER);
-    List<String> keyList = builder.tracing.propagation().keys();
-    // Use a more efficient injector if we are only propagating a single header
-    if (builder.writeB3SingleFormat || keyList.equals(Propagation.B3_SINGLE_STRING.keys())) {
```
Comment (jeqo) [inline instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java]: Really happy this is gone :)

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/TracingFilter.java [hunk 6]
```diff
@@ -32,7 +32,7 @@
   TracingFilter(KafkaStreamsTracing tracing, String spanName,
     Predicate<K, V> delegatePredicate, boolean filterNot) {
     this.kafkaStreamsTracing = tracing;
```
Comment (jeqo) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/TracingFilter.java]: nit: should we align variable names?

File: instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java (hunk 4 above)
Comment (adriancole) [reply instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java]: yep though primary sampling is only at the start of a trace

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/TracingFilter.java (hunk 6 above)
Comment (adriancole) [reply instrumentation/kafka-streams/src/main/java/brave/kafka/streams/TracingFilter.java]: sure I will sweep on that

**Pull request merged**